core/src/main/scala/epollcat/internal/ch/SocketHelpers.scala

```diff
@@ -30,8 +30,80 @@ import scala.scalanative.unsafe._
 
 private[ch] object SocketHelpers {
 
-  lazy val preferIPv4Stack =
-    java.lang.Boolean.parseBoolean(System.getProperty("java.net.preferIPv4Stack", "false"))
+  /* An approach with better software engineering becomes available
+   * when epollcat no longer supports Scala Native 0.4.n.
+   * One can eliminate the chance of races & mismatches by allocating a
+   * Scala Native socket and examining its address family. Scala Native
+   * will have done all the work.
+   *
+   * Tcp6Suite.scala uses a similar technique querying an
+   * AsynchronousServerSocketChannel.
+   */
+
+  lazy val useIPv4Stack = {
+    val systemPropertyForcesIPv4 =
+      java.lang.Boolean.parseBoolean(System.getProperty("java.net.preferIPv4Stack", "false"))
+    systemPropertyForcesIPv4 || hasOnlyIPv4Stack()
+  }
+
+  private def hasOnlyIPv4Stack(): Boolean = {
+    val addrinfo = stackalloc[Ptr[posix.netdb.addrinfo]]()
+    val hints = stackalloc[posix.netdb.addrinfo]()
+
+    hints.ai_family = posix.sys.socket.AF_INET6
+    hints.ai_flags = posix.netdb.AI_NUMERICHOST | posix.netdb.AI_NUMERICSERV
+    hints.ai_flags |= posix.netdb.AI_PASSIVE // ServerSocket
+    hints.ai_flags |= posix.netdb.AI_ADDRCONFIG
+    hints.ai_socktype = posix.sys.socket.SOCK_STREAM
+
+    val typelevelOrg6 = c"2606:50c0:8003::153"
+
+    val rtn = posix
+      .netdb
+      .getaddrinfo(
+        typelevelOrg6,
+        c"0",
+        hints,
+        addrinfo
+      )
+
+    val hasIPv6 =
+      try {
+        if (rtn == 0) {
+          // should never happen, but check anyways
+          java.util.Objects.requireNonNull(!addrinfo)
+          (!addrinfo).ai_family == posix.sys.socket.AF_INET6
+        } else {
+
+          if (rtn == posix.netdb.EAI_NONAME) { // expected on IPv4 & OK
+            false
+          } else if (rtn == posix.netdb.EAI_FAMILY) { // no IPv6 stack at all
+            false
```

Comment:
What is the difference between these cases? In what situation would we get `EAI_NONAME` instead of `EAI_FAMILY`?

Reply:
Or `EAI_ADDRFAMILY` for that matter 😄

I hope I do not regret asking, I can only imagine what you went through to identify each of these error cases!

Reply:
I am glad you asked.  Misery loves company.  This is a case of the joys of supporting at least three
operating systems (Linux, macOS, FreeBSD as a representative of all the other *BSD).
Then you have the IPv4 vs IPv6 cases.    So far I have not seen any Apple M1 issues,
count my luck.

Let me see if I can explain, without repeating the comments.  Along the way, please let me
know where I need to drop some more breadcrumbs.  This is one of those pieces of code
where I/we will surely be again, the day after forgetting the details.

EAI_ADDRFAMILY is at lease Linux & FreeBSD only.  EAI_NONAME is the POSIX specified.
Either of these says that the nameserver has no entry for the string passed in,
in this case val typelevelOrg6 = c"2606:50c0:8003::153". Since the hints specify
`posix.netdb.AI_NUMERICHOST | posix.netdb.AI_NUMERICSERV` it would
be very strange to get either of these codes, that is a "normal" situation where
we do not want to throw.

By my understanding, in order to get either of those codes, one must execute
a certain way down the network stack.


`EAI_FAMILY` is what we _really_ want, but it is hard to get (unless you
do not want it).  It says "There isn't even a network stack for that address".
Think of a Linux kernel which has been built with IPv6 support tailored off
(or disabled at boot or in /etc/sysctl.conf).  Execution never gets to the
place where it could tell if it had that address or not.

When we see this code, we know enough to force useIPv4Stack to true.

FS2 must have deal with stuff like this at the lowest levels all the time.
 
Please keep the questions coming.

Reply:
I have been testing privately on both X86_64 and macOS M1.  
If I can make a "best guess" for FreeBSD I have been taking 
it under my wing.  Unfortunately, I setting up a FreeBSD
system keeps getting priority shuffled to the back end of the
line.


Being able to run the tests on both JVM & Native is pretty handy, but
you knew that when you set it up that way.

I have been doing only short burst in kqueue, not industrial strength
pounding, but I have seen no hangs with your recent changes.  Nice work.